### PR TITLE
[Perf] dots.tts: slot-pooled batched streaming AudioVAE

### PR DESCRIPTION
**Pull request by guozhihao-224**
## Motivation
Roadmap item from #1367: move the streaming vocoder to slot-based batched state.

Upstream `VocoderStreamState` keeps scalar `total_frames` / `emitted_frames`, so `batch_size=N` is lockstep-only. Serving needs independent causal ages per request, while still decoding equal-T chunks for multiple live streams in one eager AudioVAE forward. The compiled / CUDA-Graph `stream_step` path remains disabled because of the allocator conflict with backbone decode graphs (#1392 / #1395).

**This PR lands the eager slot-pool + coalescing path only. Streaming AudioVAE CUDA Graph and first/steady/merge-step tuning stay follow-ups.**

## Modifications
- Add `DotsVocoderSlotPool`: per-slot LSTM / decoder window / frame counters; equal-T gather → one eager step → scatter; `@torch.no_grad()` on `step`/`flush`.
- Rework `DotsTTSStreamingVocoder` onto `StreamingVocoderBase` coalescing (`_can_batch_stream_chunks`): exact-T grouping via `select_step_participants` / `build_step_plan` / `run_step`; stream-done drains with flush + release under the codec lock.
- Wire `stream_slots` (default 16) through `stages.py` and `examples/configs/dots_tts.yaml`.
- Unit tests for equal-T batching, independent counters, admission failure, coalescing, flush/release.
Implementation note: the pool calls `VocoderInference` private eager helpers (`_decode_stream_latents` / `_decode_stream_window` / …) so rows can age independently. Prefer that over public `stream_step` until upstream exposes a per-slot API.
## Related
- Roadmap #1367 (streaming vocoder slot-based batched state)
- Soft file conflicts expected with #1440 (vocoder staging) and #1434 (codec lock labels); will rebase when they land
- Not claiming: streaming AudioVAE CUDA Graph; chunk / merge-step tuning; backbone `torch.compile`

## Benchmark
**Smoke (earlier, dual-GPU, n=64, generate-only, H200):** streaming c=1 flat; c=8 ~−9% RTF / +9% qps; c=16 ~−35% RTF / +43% qps / large TTFC drop. Treat as directional only (not same-card / not full EN / no WER).
**Canonical same-card A/B (in progress on this host):**
- H200 GPU2 serial baseline → slot
- Seed-TTS-Eval EN full 1088, seed 42, warmup 8, `--stream`
- Concurrency ladder `1 / 4 / 8 / 16`
- Generate then Qwen3-ASR-1.7B WER
- Will update this section with RTF / latency / TTFC percentiles, throughput, failed=0, and WER Δ when the run finishes
## Validation
- `pytest -q tests/unit_test/dots_tts/test_vocoder.py tests/unit_test/dots_tts/test_vocoder_streaming.py` — 24 passed
## Checklist
- [x] Format code with repository hooks (pending final commit)
- [x] Unit tests for slot pool + coalescing
- [x] Canonical same-card full-EN streaming A/B + WER (running)
- [ ] Rebase around #1440 / #1434 if merged first

### CODE REVIEW

**Comment by guozhihao-224** [comment]
## Benchmark (same-card A/B)
```
Hardware: NVIDIA H200 143771 MiB, same physical GPU for baseline vs slot (serial)
TTS GPU: 2 · ASR/WER GPU: 3 (Qwen3-ASR after TTS stopped)
Model: dots-studio/dots.tts-mf
Dataset: Seed-TTS EN full, n=1088 (zhaochenyang20/seed-tts-eval-arrow, --lang en)
Mode: --stream, --seed 42, --warmup 8
Concurrency ladder: c ∈ {1, 4, 8, 16}
Configs: baseline = origin/main worktree (2e607bc0); slot = this change (stream_slots: 16)
Reliability: all cells 1088/0 ok/fail; WER evaluated 1088, skipped 0
```

### Summary (slot vs baseline)

c | ΔRTF mean | Δqps | ΔTTFC mean | Δe2e latency | ΔWER corpus
-- | -- | -- | -- | -- | --
1 | +2.0% | −1.9% | +2.0% | +2.1% | −0.017 pp
4 | −1.9% | +1.8% | −3.2% | −1.8% | +0.059 pp
8 | −20.5% | +23.4% | −58.9% | −19.0% | +0.017 pp
16 | −34.9% | +46.9% | −78.2% | −31.9% | +0.075 pp


- Low concurrency (c=1/4): roughly flat (within ~2%).
- High concurrency (c=8/16): clear throughput/latency wins; TTFC improves most.
- Quality: all WER deltas within 0.10 pp budget (max +0.075 pp at c=16).
- Note: at c=16, mean inter-chunk rises (+77%) while RTF/qps/TTFC still improve — batching helps first-audio and overall RTF more than steady-state chunk spacing.

### Full speed + WER table


Config | c | ok/fail | RTF mean | RTF p95 | RTF p99 | qps | TTFC mean (s) | TTFC p95 | TTFC p99 | latency mean (s) | inter-chunk mean (s) | audio dur mean (s) | WER corpus | WER mean | WER p95
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
baseline | 1 | 1088/0 | 0.4288 | 0.5063 | 0.5762 | 0.570 | 0.338 | 0.466 | 0.512 | 1.754 | 0.1859 | 4.165 | 1.273% | 1.257% | 9.091%
baseline | 4 | 1088/0 | 0.7575 | 0.9133 | 1.0266 | 1.288 | 0.575 | 0.896 | 1.034 | 3.103 | 0.3313 | 4.173 | 1.264% | 1.240% | 9.091%
baseline | 8 | 1088/0 | 1.3774 | 1.7733 | 2.1987 | 1.447 | 2.225 | 2.915 | 3.309 | 5.523 | 0.4327 | 4.173 | 1.306% | 1.278% | 9.091%
baseline | 16 | 1088/0 | 2.8148 | 4.0411 | 5.3545 | 1.448 | 7.734 | 9.072 | 9.723 | 11.015 | 0.4305 | 4.171 | 1.256% | 1.261% | 9.091%
slot | 1 | 1088/0 | 0.4375 | 0.5249 | 0.6107 | 0.559 | 0.345 | 0.474 | 0.534 | 1.790 | 0.1896 | 4.165 | 1.256% | 1.244% | 9.091%
slot | 4 | 1088/0 | 0.7433 | 0.8858 | 1.0010 | 1.311 | 0.556 | 0.885 | 1.055 | 3.048 | 0.3258 | 4.175 | 1.323% | 1.317% | 9.091%
slot | 8 | 1088/0 | 1.0950 | 1.4094 | 1.8490 | 1.786 | 0.914 | 1.496 | 1.841 | 4.476 | 0.4670 | 4.166 | 1.323% | 1.296% | 9.091%
slot | 16 | 1088/0 | 1.8336 | 2.3462 | 3.0636 | 2.127 | 1.690 | 3.318 | 3.852 | 7.499 | 0.7622 | 4.172 | 1.331% | 1.334% | 9.091%

### Absolute deltas (slot − baseline)

c | ΔRTF | Δqps | ΔTTFC | Δlatency | Δinter-chunk | ΔWER corpus (pp)
-- | -- | -- | -- | -- | -- | --
1 | +2.03% | −1.93% | +1.98% | +2.05% | +1.99% | −0.017
4 | −1.87% | +1.79% | −3.22% | −1.77% | −1.66% | +0.059
8 | −20.50% | +23.43% | −58.90% | −18.96% | +7.93% | +0.017
16 | −34.86% | +46.89% | −78.15% | −31.92% | +77.05% | +0.075

**Comment by guozhihao-224** [comment]
@Hayden727 Rebased `perf/dots-slot-pool` onto `origin/main` (`2b45073c`), which already includes #1444 (compatible streaming AudioVAE batching) and #1440 (vocoder staging fast paths).

### Takeways
- After #1444 landed, slot-pool still shows clear gains on the same card: **~+18–22% qps**, **~−15–18% latency/RTF**, **~−40–45% TTFC**.
- Full EN (n=1088) is the stronger read: throughput lift holds up, and ITL regression shrinks from **+9.8% (n=128)** to **+1.7% (full)** — absolute **+7 ms** (0.424 → 0.432 s).
- On n=128, ITL +9.8% is **+36 ms**; mean audio per chunk is ~0.47 s on both arms, so ITL stays below per-chunk audio duration (playback can still stay ahead of real time). Primary wins remain earlier first audio and better end-to-end RTF/throughput.

### Relationship to #1444

#1444 batches through per-request `codec_state` and only coalesces when causal ages match (`total_frames` / `emitted_frames` lockstep). This PR keeps the same `StreamingVocoderBase` coalesce hooks, but stores state in a `DotsVocoderSlotPool` so equal-T requests can batch with **independent** causal ages. Non-streaming BatchVocoder (#1440 paths) is unchanged.

Conflict resolution on rebase: keep main's BatchVocoder; replace the #1444 `codec_state` streaming path with the slot-pool path.

### Setup
Hardware: NVIDIA H200, same physical GPU0 for main vs slot (serial) Model: dots-studio/dots.tts-mf Dataset: Seed-TTS EN (zhaochenyang20/seed-tts-eval-arrow, --lang en) Mode: --stream, --seed 42, --warmup 8, generate-only, c=8 Configs: main = origin/main worktree @ 2b45073c (includes #1444) slot = this branch @ f4b5790e (stream_slots: 16)

### A/B — n=128 (smoke)
Reliability: both arms **128/0** completed/failed
| Metric | main (#1444) | slot (this PR) | Δ |
|---|---:|---:|---:|
| Throughput (req/s) | 1.534 | **1.802** | **+17.5%** |
| Audio throughput (s/s) | 5.762 | **6.801** | **+18.0%** |
| Latency mean (s) | 5.149 | **4.384** | **−14.9%** |
| RTF mean | 1.446 | **1.209** | **−16.4%** |
| TTFC mean (s) | 2.562 | **1.536** | **−40.0%** |
| ITL mean (s) | 0.369 | 0.405 | +9.8% |
### A/B — full EN n=1088
Reliability: both arms **1088/0** completed/failed
| Metric | main (#1444) | slot (this PR) | Δ |
|---|---:|---:|---:|
| Throughput (req/s) | 1.444 | **1.757** | **+21.7%** |
| Audio throughput (s/s) | 6.025 | **7.327** | **+21.6%** |
| Latency mean (s) | 5.534 | **4.550** | **−17.8%** |
| RTF mean | 1.374 | **1.123** | **−18.3%** |
| TTFC mean (s) | 2.294 | **1.258** | **−45.2%** |
| ITL mean (s) | 0.424 | 0.432 | +1.7% |

### ITL note
ITL +9.8% is **+36 ms** absolute (0.369 → 0.405 s). Mean audio per chunk is ~0.47 s on both arms, so ITL remains below the per-chunk audio duration (≈0.86× chunk audio) and playback can still stay ahead of real time. The larger wins are earlier first audio and better end-to-end RTF/throughput — same qualitative tradeoff pattern as the pre-rebase full-EN c=8 cell (ITL +7.9% there).

### Follow-up (optional, not in this PR)
`select_step_participants` currently takes the full largest equal-T group. A small follow-up can cap coalesced batch size (reuse `max_batch_size` / add `stream_batch_max`) to trade a bit of throughput for lower ITL if reviewers want that knob.

File: examples/configs/dots_tts.yaml [hunk 1]
```diff
@@ -19,5 +19,6 @@ runtime_overrides:
       cuda_graph_max_bs: 16
   vocoder:
     optimize: true
+    stream_slots: 16
```
Comment (luojiaxuan) [inline examples/configs/dots_tts.yaml]: Can stream_slots derive from latent_engine max_running_requests or fail at startup when they differ? Otherwise a concurrency override can hit vocoder admission after serving starts.

File: sglang_omni/models/dots_tts/vocoder.py [hunk 2]
```diff
@@ -158,19 +160,24 @@ def __init__(
...
+            raise ValueError("dots.tts vocoder stream_slots must be positive")
         self.codec = codec
         self.optimize = bool(optimize)
         self.merge_steps = int(merge_steps) if optimize else 1
-        self._stream_chunk_batch_max = int(max_batch_size)
-        self._logged_stream_batch = False
+        self.stream_slots = int(stream_slots)
         self._batch_vocoder = DotsTTSBatchVocoder(codec)
+        self._slot_pool = slot_pool
+        self._stream_chunk_batch_max = self.stream_slots
```
Comment (luojiaxuan) [inline sglang_omni/models/dots_tts/vocoder.py]: Please keep max_batch_size as the streaming batch cap, or add an explicit stream_batch_max. This currently changes the configured cap from 4 to 16.

File: sglang_omni/models/dots_tts/vocoder.py [hunk 3]
```diff
@@ -371,5 +289,111 @@ def fallback_full_decode(
...
+            if state.slot is None:
+                continue
+            waveform = decoded[state.slot]
+            if waveform.numel():
+                out[request_id] = waveform
+        return out
+
+    def _get_or_create_pool_locked(self) -> DotsVocoderSlotPool:
+        # Caller must hold self.codec.lock (RLock).
+        if self._slot_pool is None:
```
Comment (luojiaxuan) [inline sglang_omni/models/dots_tts/vocoder.py:369]: Can we build the slot pool during executor setup instead of the first live chunk? Lazy allocation moves setup latency and OOM into the request path after readiness.

File: tests/unit_test/dots_tts/test_vocoder_streaming.py [hunk 4]
```diff
@@ -1,113 +1,309 @@
...
-    vocoder.decode_delta("req", state, is_final=False)
+def test_slot_pool_batches_equal_t_and_preserves_independent_counters() -> None:
+    inference = _FakeInference()
+    pool = DotsVocoderSlotPool(inference, num_slots=4, chunk_size=6)
+    s0 = pool.acquire()
+    s1 = pool.acquire()
+    older = torch.ones(1, 3, 5)
+    newer = torch.full((1, 3, 5), 2.0)
+
+    # Age s0 alone so total_frames diverge before the shared step.
```
Comment (luojiaxuan) [inline tests/unit_test/dots_tts/test_vocoder_streaming.py]: Please prefix this and the other new test comment block with # note (guozhihao-224):.

**Review by luojiaxuan — COMMENTED** [review]
Main concerns are streaming batch-cap semantics and startup allocation.

File: tests/unit_test/dots_tts/test_vocoder_streaming.py (hunk 4 above)
Comment (guozhihao-224) [reply tests/unit_test/dots_tts/test_vocoder_streaming.py]: fixed in 7fdeefc57827b03c81a52ef9e239382471550f95

**Comment by guozhihao-224** [comment]
### Benchmark update @ 7fdeefc
#### Smoke n=128 (128/0 both arms)
| Metric | origin/main  @c6980be8| slot (this PR) | Δ |
|---|---:|---:|---:|
| Throughput (req/s) | 1.535 | **1.690** | **+10.1%** |
| Audio throughput (s/s) | 5.783 | **6.356** | **+9.9%** |
| Latency mean (s) | 5.142 | **4.675** | **−9.1%** |
| RTF mean | 1.444 | **1.304** | **−9.7%** |
| TTFC mean (s) | 2.576 | **1.877** | **−27.1%** |
| ITL mean (s) | 0.365 | 0.399 | +9.3% |
#### Full EN n=1088 (1088/0 both arms)
| Metric | origin/main @c6980be8 | slot (this PR) | Δ |
|---|---:|---:|---:|
| Throughput (req/s) | 1.501 | **1.671** | **+11.3%** |
| Audio throughput (s/s) | 6.254 | **6.967** | **+11.4%** |
| Latency mean (s) | 5.326 | **4.784** | **−10.2%** |
| RTF mean | 1.334 | **1.188** | **−10.9%** |
| TTFC mean (s) | 2.340 | **1.663** | **−28.9%** |
| ITL mean (s) | 0.392 | 0.410 | +4.5% |

File: sglang_omni/models/dots_tts/vocoder.py (hunk 3 above)
Comment (guozhihao-224) [reply sglang_omni/models/dots_tts/vocoder.py:369]: fixed in 7fdeefc57827b03c81a52ef9e239382471550f95.`create_vocoder_executor` now calls `ensure_slot_pool()` so pool allocation / OOM happens at setup, not on the first live chunk.

File: sglang_omni/models/dots_tts/vocoder.py (hunk 2 above)
Comment (guozhihao-224) [reply sglang_omni/models/dots_tts/vocoder.py]: fixed in 7fdeefc57827b03c81a52ef9e239382471550f95,  Restored coalesce width to `max_batch_size` (not `stream_slots`).  
   `_stream_chunk_batch_max = int(max_batch_size)`, and `select_step_participants` caps with that value.  
   `stream_slots` remains admission capacity only.

File: examples/configs/dots_tts.yaml (hunk 1 above)
Comment (guozhihao-224) [reply examples/configs/dots_tts.yaml]: fixed in 7fdeefc57827b03c81a52ef9e239382471550f95, `stream_slots` is derived from `latent_engine` `max_running_requests` at startup when unset. If explicitly set and unequal, config construction fails. Example yaml no longer duplicates `stream_slots: 16`.

**Review by luojiaxuan — APPROVED** [review]
